### PR TITLE
Fix implicitly nullable arguments.

### PR DESCRIPTION
**Pull request by EreMaijala**
This is part one of improving PHP 8.4 compatibility.

### CODE REVIEW

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php [hunk 1]
```diff
@@ -882,15 +882,15 @@ protected function getBoundWithRecords($item)
     /**
      * This method queries the ILS for holding information.
      *
-     * @param string $bibId   Bib-level id
-     * @param array  $patron  Patron login information from $this->patronLogin
-     * @param array  $options Extra options (not currently used)
+     * @param string   $bibId   Bib-level id
+     * @param array  ? $patron  Patron login information from $this->patronLogin
+     * @param array    $options Extra options (not currently used)
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: This got formatted incorrectly for some reason; fix:

```suggestion
     * @param string $bibId   Bib-level id
     * @param ?array $patron  Patron login information from $this->patronLogin
     * @param array  $options Extra options (not currently used)
```

File: module/VuFind/src/VuFind/Record/Loader.php [hunk 2]
```diff
@@ -121,7 +121,7 @@ public function load(
         $id,
         $source = DEFAULT_SEARCH_BACKEND,
         $tolerateMissing = false,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Record/Loader.php:124]: Comment above needs to be adjusted to match.

File: module/VuFind/src/VuFind/Record/Loader.php [hunk 3]
```diff
@@ -204,7 +204,7 @@ public function loadBatchForSource(
         $ids,
         $source = DEFAULT_SEARCH_BACKEND,
         $tolerateBackendExceptions = false,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Record/Loader.php:207]: Same here.

File: module/VuFind/src/VuFind/Reserves/CsvReader.php [hunk 4]
```diff
@@ -162,7 +162,7 @@ protected function loadFile($fn)
             throw new \Exception("Could not open $fn!");
         }
         $lineNo = $goodLines = 0;
-        while ($line = fgetcsv($fh, 0, $this->delimiter)) {
+        while ($line = fgetcsv($fh, 0, $this->delimiter, escape: '\\')) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Reserves/CsvReader.php]: This appears to be an accidental change unrelated to the rest of this PR.

```suggestion
        while ($line = fgetcsv($fh, 0, $this->delimiter)) {
```

Comment (demiankatz) [inline module/VuFind/src/VuFind/Reserves/CsvReader.php]: This appears to be an accidental change unrelated to the rest of this PR.

```suggestion
        while ($line = fgetcsv($fh, 0, $this->delimiter)) {
```

File: module/VuFind/src/VuFind/Search/Solr/InjectSpellingListener.php [hunk 5]
```diff
@@ -85,7 +85,7 @@ class InjectSpellingListener
     public function __construct(
         BackendInterface $backend,
         array $dictionaries,
-        LoggerInterface $logger = null
+        ?LoggerInterface $logger = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Search/Solr/InjectSpellingListener.php:88]: Comment needs to be adjusted to match.

File: module/VuFind/src/VuFind/Solr/Writer.php [hunk 6]
```diff
@@ -155,7 +155,7 @@ public function save(
         $backend,
         DocumentInterface $doc,
         $handler = 'update',
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Solr/Writer.php:158]: Comment needs adjustment.

File: module/VuFind/src/VuFindTest/Container/MockContainer.php [hunk 7]
```diff
@@ -52,7 +52,7 @@ class MockContainer implements ServiceLocatorInterface
      *
      * @return mixed
      */
-    public function build($name, array $options = null)
+    public function build($name, ?array $options = null)
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Container/MockContainer.php:55]: Comment needs adjustment.

File: module/VuFind/src/VuFindTest/Feature/MockSearchCommandTrait.php [hunk 8]
```diff
@@ -54,7 +54,7 @@ trait MockSearchCommandTrait
      * @return Command
      */
     protected function getMockSearchCommand(
-        ParamBag $params = null,
+        ?ParamBag $params = null,
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Feature/MockSearchCommandTrait.php:57]: Comment needs adjustment.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php [hunk 9]
```diff
@@ -845,7 +845,7 @@ protected function assertWithTimeout(
         callable $callback,
         callable $compareFunc,
         callable $assertion,
-        int $timeout = null
+        ?int $timeout = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:848]: Comment needs adjustment.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php [hunk 10]
```diff
@@ -908,7 +908,7 @@ function ($expected, $result): bool {
     protected function assertStringContainsStringWithTimeout(
         string $expected,
         callable $callback,
-        int $timeout = null
+        ?int $timeout = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:911]: ...and here.

File: module/VuFind/src/VuFindTest/Integration/MinkTestCase.php [hunk 11]
```diff
@@ -973,7 +973,7 @@ protected function submitSearchForm(
      */
     protected function waitForPageLoad(
         Element $page,
-        int $timeout = null
+        ?int $timeout = null
```
Comment (demiankatz) [inline module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:976]: ...and here.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchLimitTest.php [hunk 12]
```diff
@@ -67,7 +67,7 @@ class SearchLimitTest extends \VuFindTest\Integration\MinkTestCase
      */
     protected function setUpLimitedSearch(
         string $limitParam,
-        string $options = null,
+        ?string $options = null,
```
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchLimitTest.php:70]: Comment needs adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Auth/ILSAuthenticatorTest.php [hunk 13]
```diff
@@ -251,9 +251,9 @@ protected function getAuthConfig(): array
      * @return ILSAuthenticator
      */
     protected function getAuthenticator(
-        Manager $manager = null,
-        ILSConnection $connection = null,
-        EmailAuthenticator $emailAuth = null,
+        ?Manager $manager = null,
+        ?ILSConnection $connection = null,
+        ?EmailAuthenticator $emailAuth = null,
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Auth/ILSAuthenticatorTest.php:256]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Recommend/SideFacetsTest.php [hunk 14]
```diff
@@ -372,10 +372,10 @@ public function testDynamicCheckboxesCanBeDisabled(): void
      * @return SideFacets
      */
     protected function getSideFacets(
-        \VuFind\Config\PluginManager $configLoader = null,
-        Results $results = null,
+        ?\VuFind\Config\PluginManager $configLoader = null,
+        ?Results $results = null,
         string $settings = '',
-        \Laminas\Stdlib\Parameters $request = null
+        ?\Laminas\Stdlib\Parameters $request = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Recommend/SideFacetsTest.php:378]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Record/LoaderTest.php [hunk 15]
```diff
@@ -338,9 +338,9 @@ protected function getDriver(string $id = 'test', string $source = 'Solr'): Mock
      */
     protected function getLoader(
         SearchService $service,
-        RecordFactory $factory = null,
-        Cache $recordCache = null,
-        FallbackLoader $fallbackLoader = null
+        ?RecordFactory $factory = null,
+        ?Cache $recordCache = null,
+        ?FallbackLoader $fallbackLoader = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Record/LoaderTest.php:343]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/RecordDriver/SolrOverdriveTest.php [hunk 16]
```diff
@@ -243,9 +243,9 @@ public function testGetAllSubjectHeadingsNonMarc(): void
      * @return SolrOverdrive
      */
     protected function getDriver(
-        Config $config = null,
-        Config $recordConfig = null,
-        OverdriveConnector $connector = null
+        ?Config $config = null,
+        ?Config $recordConfig = null,
+        ?OverdriveConnector $connector = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/RecordDriver/SolrOverdriveTest.php:248]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/RecordTab/TabManagerTest.php [hunk 17]
```diff
@@ -73,8 +73,8 @@ class TabManagerTest extends \PHPUnit\Framework\TestCase
      * @return TabManager
      */
     protected function getTabManager(
-        PluginManager $pluginManager = null,
-        ConfigManager $configManager = null
+        ?PluginManager $pluginManager = null,
+        ?ConfigManager $configManager = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/RecordTab/TabManagerTest.php:77]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/EDS/ParamsTest.php [hunk 18]
```diff
@@ -97,8 +97,8 @@ public function testDynamicCheckboxes()
      * @return Params
      */
     protected function getParams(
-        Options $options = null,
-        PluginManager $mockConfig = null
+        ?Options $options = null,
+        ?PluginManager $mockConfig = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/EDS/ParamsTest.php:101]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/HistoryTest.php [hunk 19]
```diff
@@ -162,9 +162,9 @@ public function testPurgeHistory(): void
      * @return History
      */
     protected function getHistory(
-        SearchServiceInterface $searchService = null,
-        ResultsManager $resultsManager = null,
-        \VuFind\Config\Config $config = null
+        ?SearchServiceInterface $searchService = null,
+        ?ResultsManager $resultsManager = null,
+        ?\VuFind\Config\Config $config = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/HistoryTest.php:167]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ParamsTest.php [hunk 20]
```diff
@@ -212,8 +212,8 @@ public function testSortTieBreakerParameter(
      * @return Params
      */
     protected function getParams(
-        Options $options = null,
-        PluginManager $mockConfig = null
+        ?Options $options = null,
+        ?PluginManager $mockConfig = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ParamsTest.php:216]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ResultsTest.php [hunk 21]
```diff
@@ -454,9 +454,9 @@ public function testMissingSorter(): void
      * @return Results
      */
     protected function getResults(
-        Params $params = null,
-        SearchService $searchService = null,
-        Loader $loader = null
+        ?Params $params = null,
+        ?SearchService $searchService = null,
+        ?Loader $loader = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ResultsTest.php:459]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ResultsTest.php [hunk 22]
```diff
@@ -515,8 +515,8 @@ protected function getResultsFromResponse(
      * @return Params
      */
     protected function getParams(
-        Options $options = null,
-        PluginManager $mockConfig = null
+        ?Options $options = null,
+        ?PluginManager $mockConfig = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/Solr/ResultsTest.php:519]: ...and here.

File: module/VuFind/tests/unit-tests/src/VuFindTest/Search/Summon/ParamsTest.php [hunk 23]
```diff
@@ -103,8 +103,8 @@ public function testCheckboxVisibility()
      * @return Params
      */
     protected function getParams(
-        Options $options = null,
-        PluginManager $mockConfig = null
+        ?Options $options = null,
+        ?PluginManager $mockConfig = null
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Search/Summon/ParamsTest.php:107]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/IconTest.php [hunk 24]
```diff
@@ -123,8 +123,8 @@ protected function getMockImageLink(string $expected): ImageLink
      * @return Icon
      */
     protected function getIconHelper(
-        array $config = null,
-        StorageInterface $cache = null,
+        ?array $config = null,
+        ?StorageInterface $cache = null,
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/IconTest.php:127]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordTest.php [hunk 25]
```diff
@@ -772,7 +772,7 @@ public function testGetUrlList(): void
     protected function getRecord(
         RecordDriver $driver,
         array|Config $config = [],
-        Context $context = null,
+        ?Context $context = null,
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordTest.php:775]: Comments need adjustment.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordTest.php [hunk 26]
```diff
@@ -918,9 +918,9 @@ protected function setSuccessTemplate(
      */
     protected function getCoverLoader(
         array $config = [],
-        \VuFind\Content\Covers\PluginManager $manager = null,
-        ThemeInfo $theme = null,
-        \VuFindHttp\HttpService $httpService = null,
+        ?\VuFind\Content\Covers\PluginManager $manager = null,
+        ?ThemeInfo $theme = null,
+        ?\VuFindHttp\HttpService $httpService = null,
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/RecordTest.php:923]: ...and here.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Import/WebCrawlCommandTest.php [hunk 27]
```diff
@@ -199,9 +199,9 @@ public function testForcedReadFromCache(): void
      * @return MockObject&WebCrawlCommand
      */
     protected function getMockCommand(
-        Importer $importer = null,
-        Writer $solr = null,
-        Config $config = null,
+        ?Importer $importer = null,
+        ?Writer $solr = null,
+        ?Config $config = null,
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Import/WebCrawlCommandTest.php:204]: Comment needs adjustment.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/AddUsingTemplateCommandTest.php [hunk 28]
```diff
@@ -126,8 +126,8 @@ public function testSuccessWithMinimalParameters()
      * @return AddUsingTemplateCommand
      */
     protected function getMockCommand(
-        ExtendedIniNormalizer $normalizer = null,
-        ExtendedIniReader $reader = null,
+        ?ExtendedIniNormalizer $normalizer = null,
+        ?ExtendedIniReader $reader = null,
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/AddUsingTemplateCommandTest.php:130]: Comment needs adjustment.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/CopyStringCommandTest.php [hunk 29]
```diff
@@ -206,8 +206,8 @@ public function testFailureWithMissingTextDomain()
      * @return CopyStringCommand
      */
     protected function getMockCommand(
-        ExtendedIniNormalizer $normalizer = null,
-        ExtendedIniReader $reader = null,
+        ?ExtendedIniNormalizer $normalizer = null,
+        ?ExtendedIniReader $reader = null,
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/CopyStringCommandTest.php:210]: Comment needs adjustment.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/DeleteCommandTest.php [hunk 30]
```diff
@@ -136,8 +136,8 @@ public function testDeletingNonExistentString()
      * @return AddUsingTemplateCommand
      */
     protected function getMockCommand(
-        ExtendedIniNormalizer $normalizer = null,
-        ExtendedIniReader $reader = null,
+        ?ExtendedIniNormalizer $normalizer = null,
+        ?ExtendedIniReader $reader = null,
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Language/DeleteCommandTest.php:140]: Comment needs adjustment.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Util/CreateHierarchyTreesCommandTest.php [hunk 31]
```diff
@@ -164,8 +164,8 @@ protected function getMockResultsManager($results = null)
      * @return SuppressedCommand
      */
     protected function getCommand(
-        Loader $loader = null,
-        PluginManager $results = null
+        ?Loader $loader = null,
+        ?PluginManager $results = null
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/Util/CreateHierarchyTreesCommandTest.php:168]: Comment needs adjustment.

File: module/VuFindSearch/src/VuFindSearch/Backend/BackendInterface.php [hunk 32]
```diff
@@ -73,7 +73,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/BackendInterface.php:76]: Comment needs adjustment.

File: module/VuFindSearch/src/VuFindSearch/Backend/BackendInterface.php [hunk 33]
```diff
@@ -84,5 +84,5 @@ public function search(
      *
      * @return \VuFindSearch\Response\RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null);
+    public function retrieve($id, ?ParamBag $params = null);
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/BackendInterface.php:87]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/Blender/Backend.php [hunk 34]
```diff
@@ -151,7 +151,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Blender/Backend.php:154]: Comment needs adjustment.

File: module/VuFindSearch/src/VuFindSearch/Backend/Blender/Backend.php [hunk 35]
```diff
@@ -371,7 +371,7 @@ protected function fillMergedCollection(
      *
      * @return \VuFindSearch\Response\RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Blender/Backend.php:374]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/BrowZine/Backend.php [hunk 36]
```diff
@@ -98,7 +97,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/BrowZine/Backend.php:100]: Comment needs adjustment.

File: module/VuFindSearch/src/VuFindSearch/Backend/BrowZine/Backend.php [hunk 37]
```diff
@@ -138,7 +137,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/BrowZine/Backend.php:140]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/EDS/Backend.php [hunk 38]
```diff
@@ -199,7 +199,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EDS/Backend.php:202]: Comment needs adjustment.

File: module/VuFindSearch/src/VuFindSearch/Backend/EDS/Backend.php [hunk 39]
```diff
@@ -297,7 +297,7 @@ public function search(
      *
      * @return \VuFindSearch\Response\RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EDS/Backend.php:300]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php [hunk 40]
```diff
@@ -63,8 +63,8 @@ class Backend extends AbstractBackend
     /**
      * Constructor.
      *
-     * @param Connector                        $connector EIT connector
-     * @param RecordCollectionFactoryInterface $factory   Record collection factory
+     * @param Connector                         $connector EIT connector
+     * @param ?RecordCollectionFactoryInterface $factory   Record collection factory (null for default)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php]: Code below has not been updated. Is this change actually needed here?

File: module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php [hunk 41]
```diff
@@ -90,7 +90,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php:93]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php [hunk 42]
```diff
@@ -111,7 +111,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EIT/Backend.php:114]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/EIT/Connector.php [hunk 43]
```diff
@@ -200,7 +200,7 @@ protected function call($method = 'GET', $params = null)
      * @throws \Exception
      * @return array
      */
-    public function getRecord($id, ParamBag $params = null)
+    public function getRecord($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/EIT/Connector.php:203]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/Exception/DeepPagingException.php [hunk 44]
```diff
@@ -64,7 +64,7 @@ public function __construct(
         $message,
         $code,
         $legalPage = 0,
-        Exception $prev = null
+        ?Exception $prev = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Exception/DeepPagingException.php:67]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/LibGuides/Backend.php [hunk 45]
```diff
@@ -107,7 +106,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/LibGuides/Backend.php:109]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/LibGuides/Backend.php [hunk 46]
```diff
@@ -143,7 +142,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/LibGuides/Backend.php:145]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/Pazpar2/Backend.php [hunk 47]
```diff
@@ -135,7 +135,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Pazpar2/Backend.php:138]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/Pazpar2/Backend.php [hunk 48]
```diff
@@ -184,7 +184,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Pazpar2/Backend.php:187]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/Primo/Backend.php [hunk 49]
```diff
@@ -96,7 +95,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Primo/Backend.php:98]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/Primo/Backend.php [hunk 50]
```diff
@@ -134,7 +133,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Primo/Backend.php:136]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/Solr/Backend.php [hunk 51]
```diff
@@ -133,7 +133,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Solr/Backend.php:136]: Comment needs to be updated (here and in most changes below as well).

File: module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php [hunk 52]
```diff
@@ -199,7 +199,7 @@ public function resetLastUrl()
      *
      * @return string
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php:202]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php [hunk 53]
```diff
@@ -281,7 +281,7 @@ public function terms(ParamBag $params)
     public function write(
         DocumentInterface $document,
         $handler = 'update',
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php:284]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Backend/Summon/Backend.php [hunk 54]
```diff
@@ -103,7 +102,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/Summon/Backend.php:105]: Comment needs to be updated (here and in two places below).

File: module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Backend.php [hunk 55]
```diff
@@ -94,7 +93,7 @@ public function search(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Backend.php:96]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Backend.php [hunk 56]
```diff
@@ -114,7 +113,7 @@ public function search(
      *
      * @return RecordCollectionInterface
      */
-    public function retrieve($id, ParamBag $params = null)
+    public function retrieve($id, ?ParamBag $params = null)
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Backend.php:116]: ...and here.

File: module/VuFindSearch/src/VuFindSearch/Feature/GetIdsInterface.php [hunk 57]
```diff
@@ -58,6 +58,6 @@ public function getIds(
         AbstractQuery $query,
         $offset,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Feature/GetIdsInterface.php:61]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Feature/RandomInterface.php [hunk 58]
```diff
@@ -55,6 +55,6 @@ interface RandomInterface
     public function random(
         AbstractQuery $query,
         $limit,
-        ParamBag $params = null
+        ?ParamBag $params = null
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Feature/RandomInterface.php:58]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Feature/RetrieveBatchInterface.php [hunk 59]
```diff
@@ -50,5 +50,5 @@ interface RetrieveBatchInterface
      *
      * @return \VuFindSearch\Response\RecordCollectionInterface
      */
-    public function retrieveBatch($ids, ParamBag $params = null);
+    public function retrieveBatch($ids, ?ParamBag $params = null);
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Feature/RetrieveBatchInterface.php:53]: Comment needs to be updated.

File: module/VuFindSearch/src/VuFindSearch/Feature/SimilarInterface.php [hunk 60]
```diff
@@ -50,5 +50,5 @@ interface SimilarInterface
      *
      * @return \VuFindSearch\Response\RecordCollectionInterface
      */
-    public function similar($id, ParamBag $params = null);
+    public function similar($id, ?ParamBag $params = null);
```
Comment (demiankatz) [inline module/VuFindSearch/src/VuFindSearch/Feature/SimilarInterface.php:53]: Comment needs to be updated.

File: module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Blender/BackendTest.php [hunk 61]
```diff
@@ -1050,7 +1050,7 @@ protected function getBackendForFacetsAndErrors($facets, $errors)
      */
     protected function getSearchParams(
         array $filters,
-        Query $query = null
+        ?Query $query = null
```
Comment (demiankatz) [inline module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Blender/BackendTest.php:1053]: Comment needs to be updated.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the monumental effort, @EreMaijala. I don't think it's possible to edit 575 files without missing a few details, though; see below for some notes. I've tried to use the GitHub suggest feature where possible (sadly, not as many places as I would like), so hopefully you can make at least a few fixes by just clicking a button instead of editing yourself.

In any case, all tests are passing for me, and the vast majority of these issues are just related to comments (though there are a couple that may be more significant).

**Comment by EreMaijala** [comment]
@demiankatz Thanks for thorough checks! I believe I've now addresses the doc comments listed as well as some others I noticed. I'm sure there are still issues with doc comments in methods untouched here, but I couldn't find a great way to ensure consistency.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala! In double-checking everything, I found two more minor mismatched comments, which I just went ahead and fixed. Should be safe to merge this now!

**Pull request merged**